### ISSUE TITLE
Parser: (re)allow duplicate group names (e.g. move the check out of the parser) 

### ISSUE THREAD

**Issue by hg42** [DmitrySoshnikov/regexp-tree#213]
I am currently revisiting my uxregexp extended regexp module after quite a long time. I use the regexp-tree to parse the extended regexp. However I noticed, you reject duplicate group names now, which breaks my algorithm.

Please see issue #142, where you forced group names to be unique.

As an old perl user, I am not sure, if the mentioned proposal at https://tc39.es/proposal-regexp-named-groups/#sec-patterns-static-semantics-early-errors really wants this at the end...at least, if so, I think this should change.
Perl is the mother for such extended regexps and always far ahead of other implementations, meaning there were solutions for all kinds of regexp issues when others didn't even think of these.

Example:
alternative representation of a date, e.g.
```
   # 2020-09-21
   (?<year>  [0-9]{4} ) -
   (?<month> [0-9]{2} ) -
   (?<day>   [0-9]{2} )
   |
   # 09/21/2020
   (?<month> [0-9]{2} ) /
   (?<day>   [0-9]{2} ) /
   (?<year>  [0-9]{4} )
```
so, the duplicate group name is legal and quite useful, if only one matches.

Additionally, multiple matches can be collected in arrays.
If you can collect with wildcards like:
`((?<name> some regexp) some separator regexp)*`
it is only a step further to allow multiple occurances of the same group name in one expression, e.g.:
`((?<name> some regexp) sep1 (?<name> some regexp) sep2)*`

I also think, a check shouldn't be in the parser, because a pure declarative syntax doesn't include the condition.
However a consumer of the AST would check the names as necessary.
I also think, it is better to separate such implicit rules from the parsing process.

From my POV, I see these possibilities:
* you could remove the check for a duplicate group in the parser (may be moving it to another part)
* you could make it optional (because checking at the parser is more efficient)
* if you reject this, I need to rename each duplicated group name (add a count), which introduces one more layer
**[EDIT: not really possible, I would have to change the names _before_ parsing]**

thanks for listening

PS: btw. I am not a "native" javascript developer (c++, perl, and many other languages)...
So, I wonder, what would be the best way to comment on that proposal? would they listen to my comment at all?

**Comment by DmitrySoshnikov** [comment]
@hg42 yes, I think the example you show is really useful. I'll need to double-check why ES spec chose check against duplicate group names (the use-case with alternative makes perfect practical sense), but it really should throw as a "syntax error".

This is from JS:

```
SyntaxError: Invalid regular expression: /(?<year>[0-9]{4})|(?<year>[0-9]{2})/: Duplicate capture group name
```
Maybe it's worth allowing in a special parsing mode, like `--mode perl` or something.

**Comment by hg42** [comment]
thanks for answering...

well "perl" would be a big claim...perl regexps are quite a lot more (I never got on par with all the useful features).

I think, providing some options for such simple cases would be good.
Something like option sets would be another nice feature. So, one could gather some options and set them all at once. Some sets could be part of the distribution.
I understand, having too much (especially complicated) options could create a maintenance nightmare.

**Comment by DmitrySoshnikov** [comment]
@hg42, yeah, I think this would make sense for ECMAScript itself eventually, and for now we can probably add `--loose-mode` parse option, which would allow some some features. Alternatively these could be specific [options](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/parser/index.js#L21):

```js
parser.parse(re, {
  allowGroupNameDuplicates: boolean,
});
```

**Comment by hg42** [comment]
yes, --loose-mode looks good and the way specific options are added.
I assume they would merge with the defaults?

**Comment by DmitrySoshnikov** [comment]
Here's some context on the ES spec https://github.com/tc39/proposal-regexp-named-groups/issues/44 and potentially we can fix it in the spec itself (CC @littledan, @mathiasbynens).

**Comment by DmitrySoshnikov** [comment]
Fixed in https://github.com/DmitrySoshnikov/regexp-tree/commit/e833b705681a1e7358ad6364573325841debf94a and available in `v0.1.23`.

**Comment by hg42** [comment]
thanks for the work...
And also for the pointer to the proposal, I also commented there.
